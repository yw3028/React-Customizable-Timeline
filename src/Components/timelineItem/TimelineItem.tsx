import React, { useState } from 'react';
import { STimelineItem } from '../styles';
import { useSpring, animated, config } from 'react-spring';

type ItemProps = {
  title: string;
  subtitle?: string;
  content?: string;
  img?: { url: string; alt: string };
  label?: string;
  location?: string;
  animation?: boolean;
};

const TimelineItem = ({ title, subtitle, content, img, label, location, animation }: ItemProps) => {
  const [showContent, setShowContent] = useState(false);
  const handleClick = () => {
    setShowContent(!showContent);
  };
  const itemAnimation: Object = useSpring({
    from: { transform: 'scale3d(1.25, 0.75, 1)' },
    to: { transform: 'scale3d(1, 1, 1)' },
    config: config.wobbly,
  });
  return (
    <STimelineItem>
      <animated.div style={animation ? itemAnimation : undefined} onClick={handleClick}>
        <div style={{ cursor: 'pointer' }}>
          <h4>{title}</h4>
          <h5>{subtitle}</h5>
          {img && (
            <img
              style={{ maxWidth: '400px', width: '100%', minWidth: '100px', paddingBottom: '10px' }}
              src={img.url}
              alt={img.alt}
            />
          )}
        </div>
        {showContent && (
          <div>
            <div>{content}</div>
            <div
              style={{
                fontSize: 'small',
                textShadow: '0 0 0.75rem #625261',
                padding: '10px 0',
                textAlign: 'right',
              }}
            >
              {label}
            </div>
          </div>
        )}
        {location && (
          <div style={{ fontSize: 'small', textAlign: 'right' }}>
            <span role="img" aria-label="location">
              üìç
            </span>{' '}
            {location}
          </div>
        )}
      </animated.div>
    </STimelineItem>
  );
};

export default TimelineItem;
