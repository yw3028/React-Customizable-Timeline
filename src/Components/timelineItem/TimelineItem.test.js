/* eslint-disable no-undef */
import { cleanup, render, screen } from '@testing-library/react';
import React from 'react';
import TimelineItem from './TimelineItem';
import IntersectionObserver from '../helper/testHelper';
import data from '../mocks/index';

global.IntersectionObserver = IntersectionObserver;

const renderComponent = (title, subtitle, content, img, label, location, animation) =>
  render(
    <TimelineItem
      title={title}
      subtitle={subtitle}
      content={content}
      img={img}
      label={label}
      location={location}
      animation={animation}
    />,
  );

afterEach(() => {
  cleanup();
});

describe('TimelineItem component', () => {
  it('renders correctly', async () => {
    const component = renderComponent(
      data[0].events[0].title,
      data[0].events[0].subtitle,
      data[0].events[0].content,
      data[0].events[0].img,
      data[0].events[0].label,
      data[0].events[0].location,
      data[0].events[0].animation,
    );
    await expect(component.baseElement).toMatchSnapshot();
  });

  it("Doesn't render children elements", async () => {
    render(
      <TimelineItem
        title={data[0].events[0].title}
        subtitle={data[0].events[0].subtitle}
        content={data[0].events[0].content}
        img={data[0].events[0].img}
        label={data[0].events[0].label}
        location={data[0].events[0].location}
        animation={data[0].events[0].animation}
      >
        <p>LisaPisa</p>
      </TimelineItem>,
    );

    await expect(screen.queryByText(/LisaPisa/i)).toBeFalsy();
  });

  it('render data correctly', async () => {
    renderComponent(
      data[0].events[0].title,
      data[0].events[0].subtitle,
      data[0].events[0].content,
      data[0].events[0].img,
      data[0].events[0].label,
      data[0].events[0].location,
      data[0].events[0].animation,
    );
    await expect(screen.getByText('Avengers:StartGame')).toBeTruthy();
    await expect(screen.getByText('Knowhere')).toBeTruthy();
  });
});
