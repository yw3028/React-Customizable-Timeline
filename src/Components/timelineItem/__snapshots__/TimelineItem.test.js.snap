// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineItem component renders correctly 1`] = `
<body>
  <div>
    <div
      class="sc-bxivhb SUHGw"
    >
      <div>
        <div
          style="cursor: pointer;"
        >
          <h4>
            Avengers:StartGame
          </h4>
          <h5>
            StartGame
          </h5>
          <img
            alt="picture"
            src="https://images.unsplash.com/photo-1600790078201-5490baf711d6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60"
            style="max-width: 400px; width: 100%; min-width: 100px; padding-bottom: 10px;"
          />
        </div>
        <div
          style="font-size: small; text-align: right;"
        >
          <span
            aria-label="localtion"
            role="img"
          >
            üìç
          </span>
           
          Knowhere
        </div>
      </div>
    </div>
  </div>
</body>
`;
